### PR TITLE
BOM-2015 : Unpin virtualenv 

### PR DESCRIPTION
**Pull request by mzulqarnain1**
Relevant JIRA : https://openedx.atlassian.net/browse/BOM-2015

### CODE REVIEW

**Comment by mzulqarnain1** [comment]
Hi @nedbat  and @feanil,

I am working on [BOM-2015](https://openedx.atlassian.net/browse/BOM-2015), it is about unpinning virtualenv from configuration repo and fixing the symlinks problems that come with the latest version, I have started with some small changes in this PR, and I made a [sandbox](https://tools-edx-jenkins.edx.org/job/sandboxes/job/CreateSandbox/33948/) against these changes but `createSandbox` job fails at following line https://github.com/edx/configuration/blob/master/playbooks/roles/edxapp/tasks/deploy.yml#L312
Which is strange because this line seems to use python from the virtual env directly and doesn’t seems to use symlink so it seems unrelated to changes in this PR. 

@jmbowman  thinks that 
> It's a sign that it broke codejail, and a command that was supposed to fail due to security restrictions actually worked. It's probably a result of the virtualenv's python changing from a binary to a symlink resulting in codejail now protecting the wrong executable path.

Can you shed some light on it?

**Comment by feanil** [comment]
@mzulqarnain1 I wouldn't expect the change you made to cause a codejail breakage either.  Can you link to the sandbox build so that I can review the ansible output?

**Comment by mzulqarnain1** [comment]
@feanil Link to the sandbox https://tools-edx-jenkins.edx.org/job/sandboxes/job/CreateSandbox/33948

**Comment by feanil** [comment]
@mzulqarnain1 the issue seems to be that the new sandbox venv that is created symlinks the python package to the system python.

```
feanil@virtualenvunpin4 i-043e1a4f061a275cf:/etc/apparmor.d$ ls -l /edx/app/edxapp/venvs/edxapp-sandbox/bin/python
lrwxrwxrwx 1 sandbox sandbox 18 Oct 20 10:37 /edx/app/edxapp/venvs/edxapp-sandbox/bin/python -> /usr/bin/python3.5
```
This will cause it to be outside of the app-armor rules and not properly enforce a code sandbox. You probably need to update how the sandbox venv is created in the newer version of virtualenv so that it does not do this.

**Comment by mzulqarnain1** [comment]
Hey @jmbowman I have made some progress on this task, I have used `--always-copy` [argument](https://virtualenv.pypa.io/en/latest/cli_interface.html#always-copy) of virtualenv which copies the binary file for python instead of creating a symlink towards it when creating a venv. But now when I build a sandbox following error occurs somewhere in `oauth_client_setup` 
```
15:53:35 <virtualenvunpin006.sandbox.edx.org> (0, b'', b'')
15:53:35 failed: [virtualenvunpin006.sandbox.edx.org] (item={'name': 'registrar', 'url_root': 'https://registrar-virtualenvunpin006.sandbox.edx.org', 'sso_id': 'registrar-sso-key', 'sso_secret': 'registrar-sso-secret', 'backend_service_id': 'registrar-backend-service-key', 'backend_service_secret': 'registrar-backend-service-secret', 'logout_uri': 'https://registrar-virtualenvunpin006.sandbox.edx.org/logout/', 'username': 'registrar_worker'}) => {
15:53:35     "ansible_loop_var": "item",
15:53:35     "changed": true,
15:53:35     "cmd": "/edx/bin/python.edxapp /edx/bin/manage.edxapp lms --settings=production create_dot_application --grant-type authorization-code --redirect-uris \"https://registrar-virtualenvunpin006.sandbox.edx.org/complete/edx-oauth2/\" --client-id registrar-sso-key --client-secret registrar-sso-secret --scopes user_id --skip-authorization --update registrar-sso registrar_worker\n",
15:53:35     "delta": "0:00:00.038068",
15:53:35     "end": "2020-10-27 15:53:35.669902",
15:53:35     "invocation": {
15:53:35         "module_args": {
15:53:35             "_raw_params": "/edx/bin/python.edxapp /edx/bin/manage.edxapp lms --settings=production create_dot_application --grant-type authorization-code --redirect-uris \"https://registrar-virtualenvunpin006.sandbox.edx.org/complete/edx-oauth2/\" --client-id registrar-sso-key --client-secret registrar-sso-secret --scopes user_id --skip-authorization --update registrar-sso registrar_worker\n",
15:53:35             "_uses_shell": true,
15:53:35             "argv": null,
15:53:35             "chdir": null,
15:53:35             "creates": null,
15:53:35             "executable": null,
15:53:35             "removes": null,
15:53:35             "stdin": null,
15:53:35             "stdin_add_newline": true,
15:53:35             "strip_empty_ends": true,
15:53:35             "warn": true
15:53:35         }
15:53:35     },
15:53:35     "item": {
15:53:35         "backend_service_id": "registrar-backend-service-key",
15:53:35         "backend_service_secret": "registrar-backend-service-secret",
15:53:35         "logout_uri": "https://registrar-virtualenvunpin006.sandbox.edx.org/logout/",
15:53:35         "name": "registrar",
15:53:35         "sso_id": "registrar-sso-key",
15:53:35         "sso_secret": "registrar-sso-secret",
15:53:35         "url_root": "https://registrar-virtualenvunpin006.sandbox.edx.org",
15:53:35         "username": "registrar_worker"
15:53:35     },
15:53:35     "msg": "non-zero return code",
15:53:35     "rc": 1,
15:53:35     "start": "2020-10-27 15:53:35.631834",
15:53:35     "stderr": "Traceback (most recent call last):\n  File \"/edx/bin/manage.edxapp\", line 20, in <module>\n    from safe_lxml import defuse_xml_libs\nModuleNotFoundError: No module named 'safe_lxml'",
15:53:35     "stderr_lines": [
15:53:35         "Traceback (most recent call last):",
15:53:35         "  File \"/edx/bin/manage.edxapp\", line 20, in <module>",
15:53:35         "    from safe_lxml import defuse_xml_libs",
15:53:35         "ModuleNotFoundError: No module named 'safe_lxml'"
15:53:35     ],
15:53:35     "stdout": "",
15:53:35     "stdout_lines": []
15:53:35 }
15:53:35 skipping: [virtualenvunpin006.sandbox.edx.org] => (item={'name': 'None', 'url_root': 'None', 'sso_id': 'None', 'sso_secret': 'None', 'backend_service_id': 'None', 'backend_service_secret': 'None', 'logout_uri': 'None', 'username': 'designer_worker'})  => {
15:53:35     "ansible_loop_var": "item",
15:53:35     "changed": false,
15:53:35     "item": {
15:53:35         "backend_service_id": "None",
15:53:35         "backend_service_secret": "None",
15:53:35         "logout_uri": "None",
15:53:35         "name": "None",
15:53:35         "sso_id": "None",
15:53:35         "sso_secret": "None",
15:53:35         "url_root": "None",
15:53:35         "username": "designer_worker"
15:53:35     },
15:53:35     "skip_reason": "Conditional result was False"
15:53:35 }
15:53:35 skipping: [virtualenvunpin006.sandbox.edx.org] => (item={'name': 'None', 'backend_service_id': 'None', 'backend_service_secret': 'None', 'username': 'retirement_worker'})  => {
15:53:35     "ansible_loop_var": "item",
15:53:35     "changed": false,
15:53:35     "item": {
15:53:35         "backend_service_id": "None",
15:53:35         "backend_service_secret": "None",
15:53:35         "name": "None",
15:53:35         "username": "retirement_worker"
15:53:35     },
15:53:35     "skip_reason": "Conditional result was False"
15:53:35 }
15:53:35 
15:53:35 PLAY RECAP *********************************************************************
15:53:35 virtualenvunpin006.sandbox.edx.org : ok=698  changed=438  unreachable=0    failed=1    skipped=638  rescued=0    ignored=5   
15:53:35 
15:53:36 + ret=2
15:53:36 + [[ 2 -ne 0 ]]
15:53:36 + exit 2
15:53:36 Build step 'Virtualenv Builder' marked build as failure
15:53:36 Set build name.
15:53:36 New build name is '#34187 mzulqarnain1 virtualenvunpin006'
15:53:36 Variable with name 'BUILD_DISPLAY_NAME' already exists, current value: '#34187 mzulqarnain1 virtualenvunpin006', new value: '#34187 mzulqarnain1 virtualenvunpin006'
15:53:36 [ssh-agent] Stopped.
15:53:36 Finished: FAILURE
```

But if I ssh into the sandbox and run `/edx/bin/pip.edxapp freeze` `safe_lxml` is present in installed deps like
`-e git+https://github.com/edx/edx-platform.git@b36af3930d7cae2a22d0a2300e6a1e67c0b896e1#egg=safe_lxml&subdirectory=common/lib/safe_lxml`

Any idea why this might be happening?

**Comment by jmbowman** [comment]
I suspect that my note in the ticket about symlinks to virtualenvs created by virtualenv 20+ not working holds true even if using the `--always-copy` option.  So you've gotten past the problem with codejail, and are now hitting the issue I originally expected which is the use of symlinks to the virtualenv.  I suspect you'll have to refactor the shell commands in https://github.com/edx/configuration/blob/master/playbooks/roles/oauth_client_setup/tasks/main.yml to use paths to the appropriate files in the main virtualenv, and not symlinks to them.  Basically, switch away from using the `python.edxapp` and `manage.edxapp` links.

**Comment by mzulqarnain1** [comment]
> I suspect that my note in the ticket about symlinks to virtualenvs created by virtualenv 20+ not working holds true even if using the `--always-copy` option. So you've gotten past the problem with codejail, and are now hitting the issue I originally expected which is the use of symlinks to the virtualenv. I suspect you'll have to refactor the shell commands in https://github.com/edx/configuration/blob/master/playbooks/roles/oauth_client_setup/tasks/main.yml to use paths to the appropriate files in the main virtualenv, and not symlinks to them. Basically, switch away from using the `python.edxapp` and `manage.edxapp` links.

Ok , Let me try that.

**Comment by mzulqarnain1** [comment]
Sandbox made against these changes at https://tools-edx-jenkins.edx.org/job/sandboxes/job/CreateSandbox/34363/

**Review by UsamaSadiq — APPROVED** [review]

File: playbooks/roles/edxapp/tasks/deploy.yml [hunk 1]
```diff
@@ -107,7 +107,7 @@
     - install:app-requirements
 
 - name: Create the virtualenv to install the Python requirements
-  command: "virtualenv {{ edxapp_venv_dir }} -p {{ EDXAPP_PYTHON_VERSION }}"
+  command: "virtualenv {{ edxapp_venv_dir }} -p {{ EDXAPP_PYTHON_VERSION }} --always-copy"
```
Comment (jmbowman) [inline playbooks/roles/edxapp/tasks/deploy.yml:110]: I suspect we won't actually need this change, but probably safer to do it this way for now.

**Review by jmbowman — APPROVED** [review]

**Pull request merged**